server/blob/blob-union/src/main/java/org/apache/james/blob/union/HybridBlobStore.java

```diff
@@ -128,45 +145,53 @@ private HybridBlobStore(BlobStore lowCostBlobStore, BlobStore highPerformanceBlo
+        return save(bucketName, new ByteArrayInputStream(data), storagePolicy);
     }
 
     @Override
     public Mono<BlobId> save(BucketName bucketName, InputStream data, StoragePolicy storagePolicy) {
         Preconditions.checkNotNull(data);
 
-        BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
-        return selectBlobStore(storagePolicy, Mono.fromCallable(() -> isItABigStream(bufferedInputStream)))
-            .flatMap(blobStore -> blobStore.save(bucketName, bufferedInputStream, storagePolicy));
-    }
-
-    private Mono<BlobStore> selectBlobStore(StoragePolicy storagePolicy, Mono<Boolean> largeData) {
         switch (storagePolicy) {
             case LOW_COST:
-                return Mono.just(lowCostBlobStore);
+                return lowCostBlobStore.save(bucketName, data, storagePolicy);
+
             case SIZE_BASED:
-                return largeData.map(isLarge -> {
-                    if (isLarge) {
-                        return lowCostBlobStore;
-                    }
-                    return highPerformanceBlobStore;
-                });
-            case HIGH_PERFORMANCE:
-                return Mono.just(highPerformanceBlobStore);
+                BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
+                return Mono.fromCallable(() -> isItABigStream(bufferedInputStream))
+                    .filter(Boolean::booleanValue)
+                    .flatMap(ignored -> lowCostBlobStore.save(bucketName, bufferedInputStream, storagePolicy))
+                    .switchIfEmpty(Mono.defer(() ->
+                        checkingDuplicateWriteThenSave(bucketName, readAllBytes(bufferedInputStream), storagePolicy)));
+
             default:
                 throw new RuntimeException("Unknown storage policy: " + storagePolicy);
         }
     }
 
-    private boolean isItABigStream(InputStream bufferedData) throws IOException {
+    private Mono<BlobId> checkingDuplicateWriteThenSave(BucketName bucketName, Mono<byte[]> bytesData, StoragePolicy storagePolicy) {
+        return Mono.fromCallable(configuration::isDuplicateWrite)
+            .filter(Boolean::booleanValue)
+            .flatMap(allowDuplicateWrite ->
+                bytesData.flatMap(bytes ->
+                    lowCostBlobStore.save(bucketName, bytes, storagePolicy)
+                        .then(highPerformanceBlobStore.save(bucketName, bytes, storagePolicy))))
+            .switchIfEmpty(bytesData.flatMap(bytes -> highPerformanceBlobStore.save(bucketName, bytes, storagePolicy)));
+    }
+
+    private boolean isItABigStream(BufferedInputStream bufferedData) throws IOException {
         bufferedData.mark(0);
         bufferedData.skip(configuration.getSizeThreshold());
         boolean isItABigStream = bufferedData.read() != -1;
         bufferedData.reset();
         return isItABigStream;
```

Comment:
isLargeStream might be a more concise name here.

Reply:
Agree with it